### ISSUE TITLE
[bug] Duplicated users (non-merged) after calling `init` multiples times

### ISSUE THREAD

**Issue by BrodaNoel** [amplitude/Amplitude-TypeScript#178]
## Expected Behavior
Amplitude should not duplicate users.
I think it's related to the "device id" that is different in some situations that it may not be different.

My app, in some scenarios, for anonymous users (without userId defined), may run 3 events on the startup, even before the `init` function was called.

It works something like this:

```js
onStartup() {
  analytics.track('OPEN_APP');
  analytics.track('SOME_OTHER_EVENT');
  analytics.track('GO_TO_MAIN_PAGE');
}
```

The problem there is that the `analytics.track` functions does:

```js
let wasInited = false;

const init = async () => {
    await initAmplitude(apiKey).promise;
    wasInited = true;
};

const track = async (eventName) => {
  if (!wasInited) {
    await init();
  }

  trackEvent(eventName);
};
```

So, as you can see, if we call 3 times `.track()`, in will actually run 3 times the `init` function (because there is an `await` that will release the JS thread).

It seems like this scenario was working well with the old `expo-analytics-aimplitude` library (`Amplitude-ReactNative`).
If I did that thing, in the old library, those 3 inits will actually create the same user, and will actually track the 3 events in the same user.

Currently, it's creating 3 different users with 1 even on each of them.

Check this:
<img width="1138" alt="Screen Shot 2022-08-09 at 17 41 07" src="https://user-images.githubusercontent.com/1837969/183756666-b7f6fcd5-b57b-46c5-9f2b-65df308da9e0.png">

3 different users running those 3 events with LESS THAN 1 second of difference.
And, as I can tell you, those 3 events are run always one after the other one.

These are the resulting users:

<img width="951" alt="Screen Shot 2022-08-09 at 17 42 47" src="https://user-images.githubusercontent.com/1837969/183757001-5ea53e55-2edd-42b3-a454-f5005bf03cfa.png">

<img width="944" alt="Screen Shot 2022-08-09 at 17 42 59" src="https://user-images.githubusercontent.com/1837969/183757030-a0f81019-156f-47b5-8bc2-77246674f4f5.png">

<img width="938" alt="Screen Shot 2022-08-09 at 17 43 08" src="https://user-images.githubusercontent.com/1837969/183757056-f52cc3c0-fce0-4ffb-a8db-40ebe2c13450.png">

As you can see, it's the same user IN THE APP, but different users on Amplitude.

It first create 1 user with 1 even, then 1 other user with the second event, and finally, the 3rd user with the last of the 3 events and ALSO all the remaining events that comes AFTER. As in:

Events of user 1:
<img width="460" alt="Screen Shot 2022-08-09 at 17 45 01" src="https://user-images.githubusercontent.com/1837969/183757330-761a352f-5246-406a-8cb5-8114ea36bd95.png">

Events of user 2:
<img width="471" alt="Screen Shot 2022-08-09 at 17 45 32" src="https://user-images.githubusercontent.com/1837969/183757423-1233f82f-56ff-4648-bfe7-93ba52b73799.png">

Events of user 3:
<img width="473" alt="Screen Shot 2022-08-09 at 17 45 51" src="https://user-images.githubusercontent.com/1837969/183757483-e5caf513-8541-42de-b083-d2f12a9388d4.png">

It clearly should merge those users. But the device ID seems to be different.

This is resulting in a huge increment of new users, which actually then don't exists. Those 3 users were actually only 1.
<img width="95" alt="Screen Shot 2022-08-09 at 17 46 45" src="https://user-images.githubusercontent.com/1837969/183757643-bcbcbc2e-52ec-4d48-b361-d97ccaaa54b8.png">

## Current Behavior
Not merging users

## Possible Solution
Maybe merging the users? Maybe something related to the different `DeviceID`?

## Steps to Reproduce
Check the code I added on the top.
It should be just:

```
1. Run 3 events on startup.
2. Inside the `track` function, call the `init`.
3. The `init` is going to create 3 different users, but whould have been only 1.
```

I insist that this was working well before. I have been using Amplitude since 3 years ago.

## Environment
- JS SDK Version: Expo SDK 46
- Installation Method: Yarn (`expo install`)
- Browser and Version: Actually happening in the device.

**Comment by BrodaNoel** [comment]
The workaround for this issue will be to only call `init` 1 time.
I'm implementing this... But... I want to repeat that this issue was not present in the old library.

**Comment by kevinpagtakhan** [comment]
Hi @BrodaNoel, thanks for choosing Amplitude. My best guess is that all the init calls race up to find and create their own device IDs hence seeing different device IDs. `init` sets up the SDK and performs read/writes to storage APIs to persist user session which include multiple async operations. Due to this, this SDK supports deferred initialization wherein you can send `track()` calls before or after `init()` calls.

In the meantime, I would suggest to avoid calling `init()` in parallel. We do support calling `init()` in sequence for cases you need to re-initialize with new configuration.

```js
let wasInited = false;

const init = () => {
  wasInited = true; // block future init immediately
  amplitude.init(apiKey);
};

const track = (eventName) => {
  if (!wasInited) {
    init(); // no need to await
  }

  amplitude.track(eventName); // can be called with or without init completing
};
```

Let me know if you need further support on this. On our end, we can block parallel executions of `init`, and re-allow when no `init` is pending state.

**Comment by BrodaNoel** [comment]
@kevinpagtakhan thanks for the answer.

I already fixed the issue in my end. Now I just run `init` on the startup, and I guess it should be all.

I reported this mainly because I am 100% sure this "bug" was not there in the old library. It was clearly creating the same DeviceId even if you call the `init` 10 times in a row, in parallel (and maybe you want to keep it in the same way, in order to create a retro-compatibility).

Feel free to close this issue if you consider no work is needed in your side.

**Comment by liuyang1520** [comment]
Hi, @BrodaNoel , thanks for reporting this! This lib is totally new and still in beta, so there might be some bugs. Also the behaviors might be different from the Expo one. We are working on the fix right now, thank you for supporting us and reporting these issues!

@kevinpagtakhan , thanks for the information!
As @kevinpagtakhan mentioned, now we support calling `track` before `init`, the events will be stored and fired after API key is fed with `init`. Closing this now.
